### PR TITLE
Fix: prevent accumulation of `SelectFields` in `PyTorchPredictor`

### PR DESCRIPTION
**Pull request by Cameronwood611**
*Issue #, if available:* #2947

*Description of changes:* Have simple check to not accumulate the same field names, resulting in a RecursionError.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Please tag this pr with at least one of these labels to make our release process faster:** bug fix

@abdulfatir @lostella Please review as this relates to changes you (co-)authored

### CODE REVIEW

**Comment by lostella** [comment]
@Cameronwood611 thanks for this; it seems like this fix is breaking the behaviour of models, so I'm not sure it's the right way to go. Is there any way you could share the model you're using for #2947, for debugging purposes? Or even a simpler, smaller dummy model that just runs into the issue. What kind of model are we talking about?

**Comment by Cameronwood611** [comment]
> @Cameronwood611 thanks for this; it seems like this fix is breaking the behaviour of models, so I'm not sure it's the right way to go. Is there any way you could share the model you're using for #2947, for debugging purposes? Or even a simpler, smaller dummy model that just runs into the issue. What kind of model are we talking about?

Yes! I updated the issue linked but I'll also post the example here --sorry about that. I noticed I had to raise the for loop above 1000 iterations (around the default recursion depth) for it to break; not sure if that's related to different model params, not loading in the model, etc. I'm not sure of the side effects this change has but maybe we could look into a certain filter or adding SelectFields into a set? Would be willing to help with this effort.

```python
import random
import numpy as np
import pandas as pd
from gluonts.torch import PyTorchPredictor
from gluonts.dataset.common import ListDataset, DataEntry
from gluonts.torch import TemporalFusionTransformerEstimator


estimator = TemporalFusionTransformerEstimator(
freq='S',
context_length=10,
prediction_length=5,
trainer_kwargs={'max_epochs': 5}
)
train = ListDataset([{'target': np.array([random.uniform(-1, 1) for _ in range(50)]), 'start': pd.Period('01-01-2023', freq='S')} for _ in range(20)], freq='S')

pred = ListDataset([{'target': np.array([random.uniform(-1, 1) for _ in range(50)]), 'start': pd.Period('01-01-2023', freq='S')} for _ in range(20)], freq='S')
model = estimator.train(train, pred)

x, y, z = np.array([x for x in range(10)]), np.array([y for y in range (10)]), np.array([z for z in range(10)]) #for example
start = pd.Period('01-01-2023', freq='S')
_input = ListDataset([{'target': x, 'start': start},
                          {'target': y, 'start': start},
                          {'target': z, 'start': start}], freq='S')
for i in range(5000): #used to be 1000 for it to break..
    pred = model.predict(_input)
    output = [sig.mean for sig in pred]  #this calls on a generator which causes the recursion error 

```

File: src/gluonts/torch/model/predictor.py [hunk 1]
```diff
@@ -73,9 +74,11 @@ def network(self) -> nn.Module:
...
-            self.input_names + self.required_fields, allow_missing=True
-        )
+        if self.provided_fields:
+            self.input_transform += SelectFields(
+                self.input_names + self.required_fields, allow_missing=True
+            )
+            self.provided_fields = True
         inference_data_loader = InferenceDataLoader(
             dataset,
             transform=self.input_transform,
```
Comment (lostella) [inline src/gluonts/torch/model/predictor.py]: @Cameronwood611 the problem is with the `+=` which updates the predictor state every time `.predict` is invoked. Solution should be to either do here

```python
input_transform = self.input_transform + SelectFields(...)
inference_data_loader = InferenceDataLoader(
    ...,
    transform=input_transform,
    ....,
)
```

or directly in the constructor 

```
self.input_transform = input_transform + SelectFields(...)
```

and remove lines 76-78 from here.

**Comment by lostella** [comment]
@Cameronwood611 thanks for the bug hunting and the fix, if you update it as suggested then it should work fine, and we'll backport it and release it soon

**Review by abdulfatir — APPROVED** [review]

**Comment by Cameronwood611** [comment]
> @Cameronwood611 thanks for the bug hunting and the fix, if you update it as suggested then it should work fine, and we'll backport it and release it soon

Thank you!
